src/Handler/Auth.hs

```diff
@@ -159,3 +159,68 @@ getSignoutR :: Handler Html
 getSignoutR = do
   deleteLoginData
   redirect HomeR
+
+getForgotR :: Handler Html
+getForgotR = do
+  redirectIfLoggedIn HomeR
+  (forgotFormWidget, _) <- generateFormPost forgotForm
+  renderForgot forgotFormWidget []
+
+postForgotR :: Handler Html
+postForgotR = do
+  redirectIfLoggedIn HomeR
+  ((result, widget), _) <- runFormPost forgotForm
+  case result of
+    FormSuccess email -> do
+      maybeUP <- runDB $ getUserByEmail email
+      case maybeUP of
+        (Just (Entity userKey _)) -> do
+          (Entity _ (Reset token _ _)) <- runDB $ do
+            deleteExistingResets userKey
+            createReset userKey
+          -- send link in email
+          $logInfo $ tokenText token
+          renderNotice "Success" ["Please check your email to reset your password."]
+        Nothing -> do
+          renderForgot widget ["This user does not exist."]
+    _ -> renderForgot widget []
+
+getResetR :: Handler Html
+getResetR = do
+  redirectIfLoggedIn HomeR
+  maybeToken <- lookupGetParam "token"
+  case maybeToken of
+    (Just tokenText) -> do
+      let token = Token tokenText
+      runDB deleteOldResets
```

Comment:
I take it this is intentional because you want success or failure of the transaction to delete expired resets to be independent of the transaction to obtain the user referenced by the current reset token?

Reply:
Yeah, this was my approach after your earlier comment here, that a failure of user lookup could rollback the delete if they were one transaction.  Not a big difference in either case, but this version matches the intended logic better.